lib/tools/system-calls.js

```diff
@@ -216,4 +217,178 @@ systemCallMethods.getPortFromEmulatorString = function (emStr) {
   return false;
 };
 
+systemCallMethods.getConnectedEmulators = async function () {
+  try {
+    log.debug("Getting connected emulators");
+    let devices = await this.getConnectedDevices();
+    let emulators = [];
+    for (let device of devices) {
+      let port = this.getPortFromEmulatorString(device.udid);
+      if (port) {
+        device.port = port;
+        emulators.push(device);
+      }
+    }
+    log.debug(`${emulators.length} emulator(s) connected`);
+    return emulators;
+  } catch (e) {
+    log.errorAndThrow(`Error getting emulators. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.setEmulatorPort = function (emPort) {
+  this.emulatorPort = emPort;
+};
+
+systemCallMethods.setDeviceId = function (deviceId) {
+  log.debug(`Setting device id to ${deviceId}`);
+  this.curDeviceId = deviceId;
+  this.adb.defaultArgs.push("-s", deviceId);
+};
+
+systemCallMethods.getRunningAVD = async function (avdName) {
+  try {
+    log.debug(`Trying to find ${avdName} emulator`);
+    let emulators = await this.getConnectedEmulators();
+    if (emulators.length < 1) {
+      log.errorAndThrow("No emulators connected");
+    }
+    for (let emulator of emulators) {
+      this.setEmulatorPort(emulator.port);
+      let runningAVDName = await this.sendTelnetCommand("avd name");
+      if (avdName === runningAVDName) {
+        log.debug(`Found emulator ${avdName} in port ${emulator.port}`);
+        this.setDeviceId(emulator.udid);
+        return emulator;
+      }
+    }
+    log.debug(`Emulator ${avdName} not running`);
+    return null;
+  } catch (e) {
+    log.errorAndThrow(`Error getting AVD. Original error: ${e.message}`);
+  }
+};
+
+systemCallMethods.getRunningAVDWithRetry = async function (avdName, timeoutMs = 20000) {
+  try {
+    let start = Date.now();
+    while ((Date.now() - start) < timeoutMs) {
+      try {
+        let runningAVD = await this.getRunningAVD(avdName.replace('@', ''));
+        if (runningAVD) {
+          return runningAVD;
+        }
+      } catch (e) {
+        // Do nothing.
+        log.info('Trying to get running AVD');
```

Comment:
this seems like a strange place for this log. since we log in `getRunningAVD`, I don't think we even need this? Could just do `catch (e) {}`, or:

```js
} catch (e) {
  log.info(`Couldn't get running AVD, will retry. Error was: ${e.message}`);
}
```